### PR TITLE
Ngwmn 2011 - Allow users to provide well registry data via bulk upload template

### PR DESCRIPTION
**Pull request by mbucknell**
Before making a pull request
----------------------------

- [X] Make sure all tests run
- [X] Update the changelog appropriately

Description
-----------
Added a custom view in the monitoring_location admin which provides a form with a single File update field. The file contents are validated and errors are displayed if appropriate. Otherwise a bulk_create is performed to create the monitoring locations and then redirected the the monitoring location's list.

The validation code that was in the MonitoringLocationModelAdmin class is now in the MonitoringLocation model so that it can be used in the bulk upload as well as the ModelAdmin instance.

I also refactored by splitting the admin into sub-modules. The modules admin/monitoring_location.py and admin/fetch_from_nwis.py are virtually the same as the code that was in admin.py. The majority of the brand new python code is in admin/bulk_upload. The tests were similarly reorganized. These modules were becoming too large and this change will allow future maintainers to more easily find the code that is responsible for a particular view.

After making a pull request
---------------------------
- [ ] If appropriate, put the link to the PR in the JIRA ticket
- [ ] Assign someone to review unless the change is trivial

### CODE REVIEW

File: wellregistry/registry/admin/__init__.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+"""
```
Comment (mbucknell) [inline wellregistry/registry/admin/__init__.py:1]: Was part of admin.py

File: wellregistry/registry/admin/fetch_from_nwis.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
+"""
```
Comment (mbucknell) [inline wellregistry/registry/admin/fetch_from_nwis.py:1]: Was part of admin.py. Now in its own module.

File: wellregistry/registry/admin/monitoring_location.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
+"""
```
Comment (mbucknell) [inline wellregistry/registry/admin/monitoring_location.py:1]: Was part of the admin.py module.

File: wellregistry/registry/models.py [hunk 4]
```diff
@@ -232,6 +233,42 @@ class MonitoringLocation(models.Model):
     class Meta:
         unique_together = (('site_no', 'agency'),)
 
+    def clean(self):
```
Comment (mbucknell) [inline wellregistry/registry/models.py:236]: This code was effectively in the admin module. Moving the code into the model clean method, allows it to be used for validation of both the admin editing forms as well as the new bulk upload feature.

File: wellregistry/registry/tests/test_models.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
+
+from decimal import Decimal
```
Comment (mbucknell) [inline wellregistry/registry/tests/test_models.py:2]: This is basically the same test code that was in test_admin to test the validations

**Comment by mbucknell** [comment]
The issues in the Codacy are either in the migrations or tests or are warning.

**Review by ajmccarthy — APPROVED** [review]
It looks good to me.  I don't have the Python/Django expertise to know if all of the coding is correct, but the code that I am familiar with looks correct, all of the pieces that were moved look like they are present in the new files.

File: CHANGELOG.md [hunk 6]
```diff
@@ -18,6 +18,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 -   Added USWDS and added NGWMN visual identification to all pages.
 -   Added validations to fields in monitoring location form
 -   Added a fetch_from_nwis view to the admin to allow site meta data to be fetched from NWIS.
+-   Added bulk uplad view to the admin to allow multiple sites data to be initialized from a csv file.
```
Comment (ayan-usgs) [inline CHANGELOG.md:21]: Nit: typo on "upload".

File: wellregistry/registry/admin/bulk_upload.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+            csv_file = request.FILES['file']
+            data_set = csv_file.read().decode('UTF-8')
+            data_stream = io.StringIO(data_set)
+            # Skip header
+            next(data_stream)
+            monitoring_locations = []
+            error_messages = []
+            row_index = 1
+            for row in csv.reader(data_stream):
+                row_index = row_index + 1
```
Comment (ayan-usgs) [inline wellregistry/registry/admin/bulk_upload.py:148]:
```suggestion
                row_index += 1
```

**Review by ayan-usgs — APPROVED** [review]
Looks good to go.

**Pull request merged**